src/dev_wasi_nn.erl

```diff
@@ -0,0 +1,324 @@
+                {ok, ExecContextId} ->
+                    % Run inference with session-specific context
+                    case dev_wasi_nn_nif:run_inference(Context, ExecContextId, binary_to_list(Prompt)) of
+                        {ok, Output} ->
+                            ?event(output, Output),
+                            {ok, #{
+                                <<"result">> => Output,
+                                <<"session-id">> => list_to_binary(SessionId)
+                            }};
+                        {error, Reason} ->
+                            ?event(dev_wasi_nn, {inference_failed, SessionId, Reason}),
+                            {error, Reason}
+                    end;
+                {error, Reason2} ->
+                    ?event(dev_wasi_nn, {session_init_failed, SessionId, Reason2}),
+                    {error, Reason2}
+            end;
+        {error, Reason3} ->
+            ?event(dev_wasi_nn, {model_load_failed, SessionId, ModelPath, Reason3}),
+            {error, Reason3}
+    end.
+
+%% @doc Download and retrieve a model by Arweave transaction ID.
+%% This function handles the complete model retrieval workflow including:
+%% - Starting the HTTP server for Arweave gateway access
+%% - Configuring local filesystem caching to avoid repeated downloads
+%% - Downloading the model from Arweave if not already cached
+%% - Resolving the local file path where the model is stored
+%%
+%% The function uses a two-tier storage strategy:
+%% 1. First checks local cache (hb_store_fs) for existing model
+%% 2. Falls back to Arweave gateway (hb_store_gateway) if not cached
+%% 3. Automatically caches downloaded models locally for future use
+%%
+%% @param TxID The Arweave transaction ID containing the model file as a binary.
+%% @returns {ok, LocalFilePath} where LocalFilePath is a string path to the 
+%%          cached model file, or {error, Reason} on failure.
+read_model_by_ID(TxID) ->
+    %% Start the HTTP server (required for gateway access)
+    hb_http_server:start_node(#{}),
+    %% Configure store with local caching for model files
+    LocalStore = #{
+        <<"store-module">> => hb_store_fs,
+        <<"name">> => <<"model-cache">>
+    },
+    Opts = #{
+        store => [
+            %% Try local cache first
+            LocalStore,
+            #{
+                <<"store-module">> => hb_store_gateway,
+                %% Cache results here
+                <<"local-store">> => LocalStore
+            }
+        ]
+    },
+    %% Attempt to read the model from cache or download from Arweave
+    case hb_cache:read(TxID, Opts) of
+        {ok, Message} ->
+            ?event(cache, {successfully_read_message_from_arweave}),
+            
+            %% Extract the data reference from the message
+            %% This could be either a link to existing cached data or binary data
+            DataLink = maps:get(<<"data">>, Message),
```

Comment:
Can we just use `hb_maps:get` directly here instead? Destructuring and reorganizing links breaks the abstractions and is likely to cause lots of pain later when new stores are introduced, etc.